pkg/packetbrokeragent/agent.go

```diff
@@ -0,0 +1,723 @@
+		return err
+	}
+	defer conn.Close()
+	ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("pba:conn:down:%s", events.NewCorrelationID()))
+
+	client := packetbroker.NewRouterForwarderDataClient(conn)
+	stream, err := client.Subscribe(ctx, &packetbroker.SubscribeForwarderRequest{
+		ForwarderNetId:    a.netID.MarshalNumber(),
+		ForwarderId:       a.clusterID,
+		ForwarderTenantId: a.tenantID,
+		Group:             a.subscriptionGroup,
+	})
+	if err != nil {
+		return err
+	}
+	logger := log.FromContext(ctx)
+	logger.Info("Subscribed as Forwarder")
+
+	downlinkCh := make(chan *packetbroker.RoutedDownlinkMessage)
+	defer close(downlinkCh)
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	var workers int32
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		select {
+		case downlinkCh <- msg:
+		default:
+			if int(atomic.LoadInt32(&workers)) < a.forwarderConfig.WorkerPool.Limit {
+				wg.Add(1)
+				atomic.AddInt32(&workers, 1)
+				go func() {
+					if err := a.handleDownlink(ctx, downlinkCh); err != nil {
+						logger.WithError(err).Warn("Forwarder subscriber stopped")
+					}
+					wg.Done()
+					atomic.AddInt32(&workers, -1)
+				}()
+			}
+			select {
+			case downlinkCh <- msg:
+			case <-time.After(workerBusyTimeout):
+				logger.Warn("Forwarder subscriber busy, drop message")
+			}
+		}
+	}
+}
+
+func (a *Agent) handleDownlink(ctx context.Context, downlinkCh <-chan *packetbroker.RoutedDownlinkMessage) error {
+	logger := log.FromContext(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(workerIdleTimeout):
+			return nil
+		case down := <-downlinkCh:
+			if down.Message == nil {
+				continue
+			}
+			ctx := events.ContextWithCorrelationID(ctx, fmt.Sprintf("pba:downlink:%s", down.Id))
+			var homeNetworkNetID types.NetID
+			homeNetworkNetID.UnmarshalNumber(down.HomeNetworkNetId)
+			ctx = log.NewContextWithFields(ctx, log.Fields(
+				"message_id", down.Id,
+				"from_home_network_net_id", homeNetworkNetID,
+				"from_home_network_tenant_id", down.HomeNetworkNetId,
+			))
+			if err := a.handleDownlinkMessage(ctx, down); err != nil {
+				logger.WithError(err).Debug("Failed to handle incoming downlink message")
+			}
+		}
+	}
+}
+
+func (a *Agent) handleDownlinkMessage(ctx context.Context, down *packetbroker.RoutedDownlinkMessage) error {
+	receivedAt := time.Now()
+	logger := log.FromContext(ctx)
+
+	for _, filler := range a.tenantContextFillers {
+		var err error
+		if ctx, err = filler(ctx, down.ForwarderTenantId); err != nil {
+			logger.WithError(err).Warn("Failed to fill context for incoming downlink message")
+			return err
+		}
+	}
+
+	ids, msg, err := fromPBDownlink(ctx, down.Message, receivedAt, a.forwarderConfig)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to convert incoming uplink message")
+		return err
+	}
+
+	req := msg.GetRequest()
+	logger = logger.WithFields(log.Fields(
```

Comment:
Let's reuse https://github.com/TheThingsNetwork/lorawan-stack/blob/8debfce391dc65c0e65fe649220389865daefd21/pkg/networkserver/downlink.go#L793-L819 here.
Maybe extract it into an `internal` package?

Reply:
Imo not worth it

Reply:
Can you at least make it consistent then?

Reply:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2225/commits/bd0d78f02c7ba01bf395bda84eeba30ca7d91a14